### ISSUE TITLE
python simplehttpserver hangs

### ISSUE THREAD

**Issue by davidmaxwaterman** [microsoft/WSL#1906]
* A brief description
'python -m SimpleHTTPServer 8000' hangs

* Expected results
should serve files normally

* Actual results (with terminal output if applicable)
The server hangs, requiring a ctrl-c to get it past whatever is blocking it.

```
Serving HTTP on 0.0.0.0 port 8000 ...
127.0.0.1 - - [13/Apr/2017 14:15:41] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:15:42] "GET /bower_components/webcomponentsjs/webcomponents-lite.min.js HTTP/1.1" 200 -
^C----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 51223)
Traceback (most recent call last):
  File "/usr/lib/python2.7/SocketServer.py", line 290, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 318, in process_request
    self.finish_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 331, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python2.7/SocketServer.py", line 652, in __init__
    self.handle()
  File "/usr/lib/python2.7/BaseHTTPServer.py", line 340, in handle
    self.handle_one_request()
  File "/usr/lib/python2.7/BaseHTTPServer.py", line 310, in handle_one_request
    self.raw_requestline = self.rfile.readline(65537)
  File "/usr/lib/python2.7/socket.py", line 480, in readline
    data = self._sock.recv(self._rbufsize)
KeyboardInterrupt
----------------------------------------
127.0.0.1 - - [13/Apr/2017 14:15:57] "GET /src/shop-app.html HTTP/1.1" 200 -



^C----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 51224)
Traceback (most recent call last):
  File "/usr/lib/python2.7/SocketServer.py", line 290, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 318, in process_request
    self.finish_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 331, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python2.7/SocketServer.py", line 652, in __init__
    self.handle()
  File "/usr/lib/python2.7/BaseHTTPServer.py", line 340, in handle
    self.handle_one_request()
  File "/usr/lib/python2.7/BaseHTTPServer.py", line 310, in handle_one_request
    self.raw_requestline = self.rfile.readline(65537)
  File "/usr/lib/python2.7/socket.py", line 480, in readline
    data = self._sock.recv(self._rbufsize)
KeyboardInterrupt
----------------------------------------
127.0.0.1 - - [13/Apr/2017 14:16:03] "GET /images/shop-icon-32.png HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:03] "GET /src/shop-category-data.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:03] "GET /src/shop-home.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:03] "GET /src/shop-image.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:03] "GET /src/shop-button.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:03] "GET /images/mens_outerwear.jpg HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:03] "GET /images/ladies_outerwear.jpg HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:03] "GET /images/mens_tshirts.jpg HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:03] "GET /images/ladies_tshirts.jpg HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:03] "GET /src/lazy-resources.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:03] "GET /src/shop-cart-data.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:03] "GET /src/shop-analytics.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:03] "GET /src/shop-cart-modal.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:03] "GET /src/shop-icons.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /src/shop-snackbar.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /src/shop-tabs.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /src/shop-tab.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /src/shop-network-warning.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /src/shop-404-warning.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /src/shop-cart-item.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /bower_components/iron-form/iron-form.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /bower_components/paper-spinner/paper-spinner-lite.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /bower_components/app-storage/app-localstorage/app-localstorage-document.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /src/shop-select.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /src/shop-tabs-overlay.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /src/shop-ripple-container.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /bower_components/paper-spinner/paper-spinner-behavior.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /bower_components/paper-spinner/paper-spinner-styles.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /bower_components/app-storage/app-storage-behavior.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /bower_components/promise-polyfill/promise-polyfill.html HTTP/1.1" 200 -
127.0.0.1 - - [13/Apr/2017 14:16:04] "GET /bower_components/promise-polyfill/Promise-Statics.js HTTP/1.1" 200 -
```

* Your Windows build number
15063.138

* Steps / All commands required to reproduce the error from a brand new installation 
1. install node/npm (recommend nvm) and bower (I'm using node v7.8.0 and bower 1.8.0)
1. npm install -g polymer-cli@next
1. mkdir tmp && cd tmp
1. polymer init
1. choose the shop
1. let it install
1. run `python -m SimpleHTTPServer 8000
1. open IE11 on localhost:8000

**Comment by davidmaxwaterman** [comment]
I should note that `polymer serve` (part of polymer-cli@next) works just fine.

**Comment by cheshrkat** [comment]
I've noticed a similar issue which I suspect will have the same root cause (I can push this out to a separate issue if preferred). 

My test case:

1. Choose a location to serve files
2. Run `python -m SimpleHTTPServer 8000` in that location
3. View `http://localhost:8000`
4. Attempt to close server with `ctrl+c` command

For me, the server does start without any issues and happily serves up the location from (1). However, when I get to (4) and want to shut it down with `ctlr+c` I have to hit `ctrl+c` at least twice, quite fast. Expected behaviour is to simply close the server.

I'm on Win10 15063.138 (aside - how I wish you could copy and paste that!) with WSL reinstalled fresh to Ubuntu 16 after the creator's update.

**Comment by cpintea** [comment]
I have the same issue on Ubuntu. Start server with python -m SimpleHTTPServer 8081. It will  work serving several request (between 1 to 10) and hang after that. After one or several ctrl+c it will continue to  work  for another 1 to 10 requests (i use chrome on w7  to make the requests).

**Comment by aseering** [comment]
@cheshrkat -- you can copy&paste from the console by selecting the text and pressing Enter :-)

@cpintea -- could ypu confirm that you are running the same build as @cheshrkat?  (Type `ver` at a Windows Command Prompt.)

**Comment by cheshrkat** [comment]
@aseering I'm talking about Win10's system info panel :) How do I get the windows version from the console?

**Comment by aseering** [comment]
@cheshrkat -- ah, I see :-)  You can type `ver` at a Windows Command Prompt.

**Comment by cheshrkat** [comment]
@aseering great tip, thanks! :)

**Comment by mattpr** [comment]
Pretty sure this is a python issue and not a windows issue since I have the same problem on macos.  ;)

I couldn't find an existing bug, so opened one here: https://bugs.python.org/issue31639

Issue is present in python 2 and 3.

**Comment by aseering** [comment]
@mattpr is your issue specific to Chrome, or can you reproduce it with a command-line tool like curl or wget?

I think I remember that there was an issue between Chrome and SimpleHttpServer at one point -- Chrome always made multiple concurrent HTTP requests; SimpleHttpServer is (was?) single-threaded and would only process one request at a time; in some scenarios this could lead to a deadlock.  I ran into that many years ago; not sure what has happened in the interim.

There definitely were WSL-specific issues with SimpleHttpServer at one point.  See #610 and #616 , for example.  Those were fixed a long time ago, but this ticket was filed a while ago too :-)

**Comment by mattpr** [comment]
Good idea.  Seems to hold water.

A test with wget seems to support this.

`wget -E -H -k -K -p http://domain1.com:8000/domain1.html`

...does not hang, whereas requests from chrome do hang.

On some level I guess I wonder about the usefulness of simple web servers if they choke on very basic website requests from modern browsers.

**Comment by davidmaxwaterman** [comment]
I opened this issue and I can't seem to reproduce the original problem, so I suggest this issue be closed.

**Comment by nowherenearithaca** [comment]
Note - I was seeing this as well.  I treated it for now by doing some stuff to avoid too many quick requests.  A patch went into python last week that introduces a threaded version of the simplehttpserver that might address it

**Comment by karray** [comment]
same here, but with all python scripts.
It probably happens after PC returns from standby mode.

**Comment by mattpr** [comment]
The python bug report I linked contains technical discussion about why this bug happens and there has already been a patch.   https://bugs.python.org/issue31639  It was merged into python 3.7.  Looks like no one has made a backport for any earlier versions of python however.
